multi: Make curl_multi_waitfds consistent with the documentation.

chrisd-work opened the pull request:
Issue Addressed:

 - The documentation of curl_multi_waitfds indicates that users should be able to call curl_multi_waitfds with a NULL ufds. However, before this change, the function would return CURLM_BAD_FUNCTION_ARGUMENT.
 - Additionally, the documentation suggests that users can use this function to determine the number of file descriptors (fds) needed. However, the function would stop counting fds if the supplied fds were exhausted.

Changes Made:

 - NULL ufds Handling: curl_multi_waitfds can now accept a NULL ufds if size is also zero.
 - Counting File Descriptors: If curl_multi_waitfds is passed a NULL ufds, or the size of ufds is insufficient, the output parameter fd_count will return the number of fds needed. This value may be higher than actually needed but never lower.

Testing:

 - Test 2405 has been updated to cover the usage scenarios described above.

Ref: #15146

dfandrich commented:
Analysis of PR #15155 at [25097ebb](https://github.com/curl/curl/commit/25097ebba475be4c35cb153197491ef1d08b4959):

[Test 1631 failed](https://github.com/curl/curl/actions/runs/11188020783/job/31106094130#check-step-3), which has NOT been flaky recently, so **there could be a real issue in the PR**.

[Test 1632 failed](https://github.com/curl/curl/actions/runs/11188020783/job/31106094130#check-step-3), which has NOT been flaky recently, so **there could be a real issue in the PR**.

###### Generated by [Testclutch](https://testclutch.curl.se/)

chrisd-work commented:
These tests appear to be skipped when I run locally on Ubuntu 22.04, is there something I have to install or configure to run these locally?

dfandrich commented:
Test 2405 needs HTTP/2 support which means a http/2 library must be compiled in to curl and nghttpx needs to be installed for tests.

chrisd-work commented:
@dfandrich Sorry I was able to run test 2405 no problem. I meant the tests that are failing checks, 1631 and 1632.

dfandrich commented:
1631/2 need FTP & HTTP support enabled and stunnel at run time in order to run.

jay commented:
> * Counting File Descriptors: If curl_multi_waitfds is passed a NULL ufds, or the size of ufds is insufficient, the output parameter fd_count will return the number of fds needed. This value may be higher than actually needed but never lower.

If ufds is NULL and size is 0 then the number written to fd_count should be exact. When would this not be the case?

If ufds is not NULL and the size is insufficient then [CURLM_OUT_OF_MEMORY](https://curl.se/libcurl/c/libcurl-errors.html#CURLMOUTOFMEMORY) is returned. IMO it is not explicit what fd_count receives in this case since the function has returned an error. However, I would think that it should also be the exact number of file descriptors.

chrisd-work commented:
@jay 

> If ufds is NULL and size is 0 then the number written to fd_count should be exact. When would this not be the case?

Great question.

One of things that the code can do, but I don't know how often it does it is combine sockets together.  Looking at the state of the code on the branch master,  `curl_multi_waitfds` calls `Curl_waitfds_add_ps` calls `cwfds_add_sock`.

This is the part of the code we need to look at, notably, this block of code:

```
  if(cwfds->n <= INT_MAX) {
    for(i = (int)cwfds->n - 1; i >= 0; --i) {
      if(sock == cwfds->wfds[i].fd) {
        cwfds->wfds[i].events |= events;
        return CURLE_OK;
      }
    }
  }
 ```
 
What this code does is "combine" sockets together. There are two cases. Perhaps socket 1 appears 3 times, so it is added only once. This also combines the `events` flags of all instances of a given socket. In then end it may end up with {1, CURL_POLL_IN | CURL_POLL_OUT, 0}` from several sockets.

To ensure `fd_count` is exactly correct in the `curl_multi_waitfds(muli, NULL, 0, &fd_count)` case I only have two choices. The first is to allocate an auxiliary buffer. This seems to be against the API, if the API does that, why doesn't it just return the allocated buffer in the first place, and provide an API to free it?

The second choice is to not consolidate. I think this is problematic because it becomes less poll like.

If however, you disagree and prefer one of the two above options I am happy to change it, just let me know.

jay commented:
I'd like to hear from others what they think about this. Basically the doc implies it returns the number of file descriptors in fd_count. I suppose it could be amended to say that if size is 0 then fd_count is greater than or equal to that amount:

> The client code can pass _size_ equal to zero just to get the number of the descriptors and allocate appropriate storage for them to be used in a subsequent function call. In this case, _fd_count_ is greater than or equal to the number of descriptors.

bagder commented:
>  I suppose it could be amended to say that if size is 0 then fd_count is greater than or equal to that amount:

It seems reasonable to me!

chrisd-work commented:
I added a SIGTRAP to the all the functions that I changed, and and the SIGTRAP is not encountered in either 1631, or 1632. So it seems that these are false positives. In addition, the only test (aside from the few that are skipped on my setup) that failed as a result of adding SIGTRAPs is 2405, the one I would expect to fail with a SIGTRAP.

This means the only test that is impacted by this change should be 2405. So I don't understand the previous or current failures.

chrisd-work commented:
In addition, I did try to repo the failure in 1631 and 1632 over the weekend for about 20k runs of each. Never hitting an issue.

jay commented:
I do not see any failing tests. I set the one failing CI job to rerun.

jay commented:
> > I suppose it could be amended to say that if size is 0 then fd_count is greater than or equal to that amount:
> 
> It seems reasonable to me!

Ok let's go with that then.

~~~diff
diff --git a/docs/libcurl/curl_multi_waitfds.md b/docs/libcurl/curl_multi_waitfds.md
index 0a8440e..2bb5805 100644
--- a/docs/libcurl/curl_multi_waitfds.md
+++ b/docs/libcurl/curl_multi_waitfds.md
@@ -50,7 +50,8 @@ be checked for being ready to read or write.
 
 The client code can pass *size* equal to zero just to get the number of the
 descriptors and allocate appropriate storage for them to be used in a
-subsequent function call.
+subsequent function call. In this case, *fd_count* receives a number greater
+than or equal to the number of descriptors.
 
 # %PROTOCOLS%
 ~~~



> I do not see any failing tests. I set the one failing CI job to rerun.

'Windows / msvc, CM x64-windows schannel MultiSSL U' is failing due to vcpkg issues which cannot possibly be related so you can ignore that one

bagder commented:
@chrisd-work will you absorb that edit and fix the merge conflict while at it?

chrisd-work commented:
@bagder updated @jay's diff, and rebased onto tot and resolved conflicts.

bagder commented on lib/select.c:
```diff
@@ -534,10 +535,8 @@ CURLcode Curl_waitfds_add_ps(struct curl_waitfds *cwfds,
-    if(events) {
-      if(cwfds_add_sock(cwfds, ps->sockets[i], events))
-        return CURLE_OUT_OF_MEMORY;
-    }
```
This is an error detection that properly returns an error that you remove here. It does not seem okay, We should not ignore error, that will lead to weird and inexplicable behaviors.

chrisd-work replied:
This error condition is covered, but how it works has changed.

The functions `cwfds_add_sock`, `Curl_waitfds_add_ps`, and `Curl_cpool_add_waitfds` were changed to no longer return an error, but rather return the  number of entries they needed. These functions will still ensure they do not overwrite past the end of the buffer.

The out of memory test has been moved to the following block.

```
  if(need != cwfds.n && ufds) {
    result = CURLM_OUT_OF_MEMORY;
  }
```

Notably if the need is not equal to the number of entries written out, then an `CURLM_OUT_OF_MEMORY`. It should never be the case that `need < cwfds.n`

Another option, is to retain the error code, and add a output parameter to all three functions that updates the `need`. I tried both methods when developing this pull request, and this one seemed simpler.

chrisd-work commented:
@bagder , I responded to your comment, any updates?

jay commented:
Thanks. I landed this with a slight modification to the example to demonstrate that &fd_count should be used on the second call to get the actual count.

~~~diff
diff --git a/docs/libcurl/curl_multi_waitfds.md b/docs/libcurl/curl_multi_waitfds.md
index 2bb5805..e2ebd37 100644
--- a/docs/libcurl/curl_multi_waitfds.md
+++ b/docs/libcurl/curl_multi_waitfds.md
@@ -45,7 +45,7 @@ If a number of descriptors used by the multi_handle is greater than the
 *size* parameter then libcurl returns CURLM_OUT_OF_MEMORY error.
 
 If the *fd_count* argument is not a null pointer, it points to a variable
-that on returns specifies the number of descriptors used by the multi_handle to
+that on return specifies the number of descriptors used by the multi_handle to
 be checked for being ready to read or write.
 
 The client code can pass *size* equal to zero just to get the number of the
@@ -90,7 +90,7 @@ int main(void)
     ufds = (struct curl_waitfd*)malloc(fd_count * sizeof(struct curl_waitfd));
 
     /* get wait descriptors from the transfers and put them into array. */
-    mc = curl_multi_waitfds(multi, ufds, fd_count, NULL);
+    mc = curl_multi_waitfds(multi, ufds, fd_count, &fd_count);
 
     if(mc != CURLM_OK) {
       fprintf(stderr, "curl_multi_waitfds() failed, code %d.\n", mc);
~~~

chrisd-work commented:
Thank you, I can't wait for the next release so we can remove the WAR for issues this PR fixed.

dkarpov1970 commented:
@chrisd-work  Thanks a lot for implementing this extension of `curl_multi_waitfds`!
Somehow, I completely missed this discussion.

I was planning to do the changes similar like you did  after the introduction of the `curl_multi_waitfds` (that's why the documentation mentioned using NULL), but unfortunately never had time to do that. 

So, I very appreciate your effort!
